File: tom_som_java_runtime/src/tom_som_runtime/SpecSectionId.java
package tom_som_runtime;

/**
 * Section-ID derivation for list/pattern sections (AA1 acceptance criteria 3–6)
 * — a faithful port of {@code spec_section_id.dart} / {@code spec_section_id.py}.
 *
 * <p>A single-valued section carries a <i>fixed</i> id — its {@code @SectionId},
 * which is already the path segment, so nothing is derived. A <b>list item</b>,
 * by contrast, gets a generated id built from the list field's
 * {@code @SectionIdPattern} (e.g. {@code DACEN-ITEM-xxx}):
 *
 * <pre>{@code <prefix><two-letter-date><number-within-the-day>}</pre>
 *
 * where {@code <prefix>} is the pattern with its trailing placeholder
 * ({@code xxx}) stripped, {@code <two-letter-date>} encodes the creation date
 * (see {@link #encodeTwoLetterDate}), and {@code <number-within-the-day>} is the
 * 1-based ordinal of the item among the list's items created the same day.
 *
 * <p>The within-day number is derived from the list's <i>current</i> ids as
 * {@code max(existing for that day) + 1}. This gives both required behaviours
 * with one rule (criterion 6): deleting a middle item never renumbers the rest
 * (the max is unchanged), while deleting the last item lowers the max so a new
 * same-day item <i>reuses</i> the just-freed id.
 */
public final class SpecSectionId {
  private SpecSectionId() {}

  /**
   * The reserved {@code refersTo} slot naming a registry entry's own section id
   * rather than one of its form fields ({@code tom_specs_model_rules.md} §6.2).
   *
   * <p>A registry key is written {@code <SECTIONID>.<slot>}; this is the one slot
   * that is not a form field name. Named because the instance-tier reference
   * check keys its registries by it, and it must be the same literal the static
   * tier in {@code tom_specs_clitool} accepts.
   */
  public static final String K_SECTION_ID_SLOT = "@sectionId";

  /**
   * The id a list item is <i>identified by</i> — its {@code storedId} when it has
   * one, and otherwise the positional id derived from the list's {@code pattern}
   * (YRD3).
   *
   * <p>An item acquires a stored id when it is created through the editor (an AA1
   * generated id) or when a document overrides one (criterion 5). Items authored
   * without one are <b>anonymous</b> and are identified by their 1-based
   * {@code position} — the pattern with its {@code xxx} placeholder replaced
   * ({@code GOAL-ITEM-xxx} → {@code GOAL-ITEM-1}), falling back to
   * {@code <fallbackStem>-<position>} for a pattern-less list.
   *
   * <p>Named once because two consumers must agree on it exactly: the markdown
   * writer, which emits it as the heading id, and the instance-tier reference
   * check, which resolves {@code @sectionId} references against it. A divergence
   * there would not be a cosmetic difference — it would red-flag a document whose
   * references are correct.
   */
  public static String effectiveListItemSectionId(
      String storedId, String pattern, int position, String fallbackStem) {
    if (storedId != null) {
      return storedId;
    }
    if (pattern != null) {
      return pattern.replace("xxx", String.valueOf(position));
    }
    return fallbackStem + "-" + position;
  }

  /**
   * Encodes a ({@code month}, {@code day}) pair as the two-letter day code used
   * in generated section ids (criterion 4).
   *
   * <ul>
   *   <li>First letter — month: {@code Jan → A}, {@code Feb → B}, … {@code Dec → L}.
   *   <li>Second letter — day-of-month: days {@code 1–26 → A–Z}; days
   *       {@code 27–31 → 0,1,2,3,4}.
   * </ul>
   */
  public static String encodeTwoLetterDate(int month, int day) {
    char monthLetter = (char) (0x41 + (month - 1));
    char dayCode;
    if (day <= 26) {
      dayCode = (char) (0x41 + (day - 1));
    } else {
      // 27 → '0', 28 → '1', … 31 → '4'.
      dayCode = (char) (0x30 + (day - 27));
    }
    return "" + monthLetter + dayCode;
  }

  /**
   * The static prefix of a {@code @SectionIdPattern}: the pattern with its
   * trailing placeholder (the run of trailing {@code x} characters) removed.
   * {@code DACEN-ITEM-xxx} → {@code DACEN-ITEM-}.
   */
  public static String sectionIdPatternPrefix(String pattern) {
    int end = pattern.length();
    while (end > 0 && pattern.charAt(end - 1) == 'x') {
      end--;
    }
    return pattern.substring(0, end);
  }

  /**
   * Builds the generated section id for a new list item (criteria 3–4, 6).
   *
   * <p>{@code pattern} is the list field's {@code @SectionIdPattern};
   * {@code month}/{@code day} are the creation date; {@code existingIds} are the
   * ids already assigned to the list's items. The within-day number is
   * {@code max(existing ids that share this day's prefix) + 1}.
   */
  public static String generateListItemSectionId(
      String pattern, int month, int day, Iterable<String> existingIds) {
    String dayPrefix = sectionIdPatternPrefix(pattern) + encodeTwoLetterDate(month, day);
    int maxForDay = 0;
    for (String id : existingIds) {
      if (id == null || !id.startsWith(dayPrefix)) {
        continue;
      }
      String tail = id.substring(dayPrefix.length());
      Integer n = parseNonNegativeInt(tail);
      if (n != null && n > maxForDay) {
        maxForDay = n;
      }
    }
    return dayPrefix + (maxForDay + 1);
  }

  private static Integer parseNonNegativeInt(String s) {
    if (s.isEmpty()) {
      return null;
    }
    for (int i = 0; i < s.length(); i++) {
      if (!Character.isDigit(s.charAt(i))) {
        return null;
      }
    }
    try {
      return Integer.valueOf(s);
    } catch (NumberFormatException e) {
      return null;
    }
  }
}

File: tom_som_rust_runtime/src/spec_section_id.rs
//! `spec_section_id` — section-ID derivation for list/pattern sections (AA1
//! acceptance criteria 3–6), a faithful port of the Go `spec_section_id.go` (and
//! the Dart/TypeScript/Python/Java references).
//!
//! A single-valued section carries a *fixed* id — its `@SectionId`, which is
//! already the path segment, so nothing is derived. A **list item**, by
//! contrast, gets a generated id built from the list field's `@SectionIdPattern`
//! (e.g. `DACEN-ITEM-xxx`):
//!
//!   `<prefix><two-letter-date><number-within-the-day>`
//!
//! where `<prefix>` is the pattern with its trailing placeholder (`xxx`)
//! stripped (`DACEN-ITEM-`), `<two-letter-date>` encodes the creation date (see
//! [`encode_two_letter_date`]), and `<number-within-the-day>` is the 1-based
//! ordinal of the item among the list's items created the same day.
//!
//! The within-day number is derived from the list's *current* ids as
//! `max(existing for that day) + 1`. This gives both required behaviours with
//! one rule (criterion 6): deleting a middle item never renumbers the rest (the
//! max is unchanged, so a new same-day item takes the next free number and the
//! numbering may stay non-consecutive), while deleting the last item lowers the
//! max so a new same-day item *reuses* the just-freed id.

use std::time::{SystemTime, UNIX_EPOCH};

use crate::spec_paths::is_all_digits;

/// Encodes a `(month, day)` pair as the two-letter day code used in generated
/// section ids (criterion 4).
///
///   - First letter — month: Jan → A, Feb → B, … Dec → L.
///   - Second letter — day-of-month: days 1–26 → A–Z; days 27–31 → 0,1,2,3,4.
pub fn encode_two_letter_date(month: i64, day: i64) -> String {
    let month_letter = char::from_u32((0x41 + (month - 1)) as u32).unwrap_or('?');
    let day_code = if day <= 26 {
        char::from_u32((0x41 + (day - 1)) as u32).unwrap_or('?')
    } else {
        // 27 → '0', 28 → '1', … 31 → '4'.
        char::from_u32((0x30 + (day - 27)) as u32).unwrap_or('?')
    };
    let mut out = String::with_capacity(2);
    out.push(month_letter);
    out.push(day_code);
    out
}

/// Returns the static prefix of a `@SectionIdPattern`: the pattern with its
/// trailing placeholder (the run of trailing `x` characters) removed.
/// `DACEN-ITEM-xxx` → `DACEN-ITEM-`.
pub fn section_id_pattern_prefix(pattern: &str) -> String {
    let bytes = pattern.as_bytes();
    let mut i = bytes.len();
    while i > 0 && bytes[i - 1] == b'x' {
        i -= 1;
    }
    pattern[..i].to_string()
}

/// Returned when a section id would collide with another id in the same list
/// (criterion 5: overriding an id must keep every id in the list unique).
#[derive(Debug, Clone, PartialEq)]
pub struct SpecSectionIdCollision {
    pub id: String,
    pub list_path: String,
}

impl std::fmt::Display for SpecSectionIdCollision {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "SpecSectionIdCollision: section id \"{}\" is already used in list \"{}\"; section ids within a list must be unique.",
            self.id, self.list_path
        )
    }
}

/// The error kind returned by the section-id write paths: a uniqueness collision
/// (criterion 5) or an attempt to set an id on a path that is not a live list
/// item.
#[derive(Debug, Clone, PartialEq)]
pub enum SpecSectionIdError {
    /// The requested id is already used by another item of the same list.
    Collision(SpecSectionIdCollision),
    /// The path is not (or no longer) a live list item.
    NotLiveItem { item_path: String },
}

impl std::fmt::Display for SpecSectionIdError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            SpecSectionIdError::Collision(c) => write!(f, "{}", c),
            SpecSectionIdError::NotLiveItem { item_path } => {
                write!(f, "'{}' is not a live list item", item_path)
            }
        }
    }
}

impl std::error::Error for SpecSectionIdError {}

/// Reports whether `err` is a uniqueness collision (as opposed to a not-a-live-
/// item error) — the Rust analogue of the Go conformance runner's
/// `errors.As(&SpecSectionIDCollision)`.
pub fn is_collision(err: &SpecSectionIdError) -> bool {
    matches!(err, SpecSectionIdError::Collision(_))
}

/// Builds the generated section id for a new list item (criteria 3–4, 6).
///
/// `pattern` is the list field's `@SectionIdPattern`; `month`/`day` are the
/// creation date; `existing_ids` are the ids already assigned to the list's
/// items. The within-day number is `max(existing ids that share this day's
/// prefix) + 1`.
pub fn generate_list_item_section_id(
    pattern: &str,
    month: i64,
    day: i64,
    existing_ids: &[String],
) -> String {
    let day_prefix = section_id_pattern_prefix(pattern) + &encode_two_letter_date(month, day);
    let mut max_for_day: i64 = 0;
    for id in existing_ids {
        if id.is_empty() || !id.starts_with(&day_prefix) {
            continue;
        }
        let tail = &id[day_prefix.len()..];
        if tail.is_empty() || !is_all_digits(tail) {
            continue;
        }
        if let Ok(n) = tail.parse::<i64>() {
            if n > max_for_day {
                max_for_day = n;
            }
        }
    }
    format!("{}{}", day_prefix, max_for_day + 1)
}

/// Returns today's `(month, day)` in local-agnostic UTC terms, using only the
/// standard library (the runtime carries zero external dependencies).
///
/// Used by [`crate::som_facade::SomList::add`] to date a generated id when no
/// explicit date is supplied; the exact value is never asserted in tests (only
/// list length/content are), so a UTC civil date is sufficient.
pub fn today_month_day() -> (i64, i64) {
    let secs = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_secs() as i64)
        .unwrap_or(0);
    let days = secs.div_euclid(86_400);
    let (_, month, day) = civil_from_days(days);
    (month, day)
}

/// Converts a count of days since 1970-01-01 into a `(year, month, day)` civil
/// date using Howard Hinnant's dependency-free algorithm.
fn civil_from_days(z_in: i64) -> (i64, i64, i64) {
    let z = z_in + 719_468;
    let era = if z >= 0 { z } else { z - 146_096 } / 146_097;
    let doe = z - era * 146_097; // [0, 146096]
    let yoe = (doe - doe / 1460 + doe / 36_524 - doe / 146_096) / 365; // [0, 399]
    let y = yoe + era * 400;
    let doy = doe - (365 * yoe + yoe / 4 - yoe / 100); // [0, 365]
    let mp = (5 * doy + 2) / 153; // [0, 11]
    let d = doy - (153 * mp + 2) / 5 + 1; // [1, 31]
    let m = if mp < 10 { mp + 3 } else { mp - 9 }; // [1, 12]
    (if m <= 2 { y + 1 } else { y }, m, d)
}
